File: research/adaptive_centering/plot_results.R
# Recreate the source article's scientific panels from the full BRM draws.
# Base R + cairo are sufficient; no plotting package installation is needed.
args <- commandArgs(trailingOnly = TRUE)
stopifnot(length(args) %in% c(1L, 2L))
input <- normalizePath(args[[1]], mustWork = TRUE)
online_input <- if (length(args) == 2L) normalizePath(args[[2]], mustWork = TRUE) else input
output <- file.path(input, "figures")
dir.create(output, showWarnings = FALSE)
read_table <- function(name) read.delim(file.path(input, name), check.names = FALSE)
blue <- "#0B7BEC"
colors <- c("#0B7BEC", "#E67E22", "#16877A", "#984EA3")
bases <- c(1L, 2L, 19L, 20L)
save_plot <- function(name, width, height, draw) {
  png(file.path(output, paste0(name, ".png")), width = width, height = height,
      units = "in", res = 160, type = "cairo", bg = "white")
  par(family = "sans", col.axis = "#353535", col.lab = "#252525", las = 1, bty = "l")
  draw()
  dev.off()
  message("figure: ", name)
}

save_plot("hsgp_basis", 10, 4.2, function() {
  par(mfrow = c(1, 2), mar = c(4, 4.3, 2.5, 1))
  x <- seq(-1.5, 1.5, length.out = 501)
  basis <- sapply(bases, function(j) sin(pi / 3 * (x + 1.5) * j) / sqrt(1.5))
  matplot(x, basis, type = "l", lty = 1, lwd = 1.5, col = colors,
          xlab = "Scaled time", ylab = "Basis function", main = "HSGP basis functions")
  abline(v = c(-1, 1), lty = 3, col = "grey45")
  legend("topright", paste("Basis", bases), col = colors, lty = 1, bty = "n", cex = .8)
  j <- seq(1, 20, length.out = 201)
  scales <- sapply(c(-2, -1, 0), function(r)
    exp(-.25 * (j * exp(r) * pi / 3)^2 + .45946926660233633 + .5 * r))
  matplot(j, scales, type = "l", lty = 1, lwd = 2, col = colors[1:3],
          xlab = "Basis frequency", ylab = "Prior spectral SD", main = "Squared-exponential spectrum")
  legend("topright", c("log length = -2", "log length = -1", "log length = 0"),
         col = colors[1:3], lty = 1, bty = "n", cex = .8)
})

observed <- read_table("observations.tsv")
posterior <- function(label, title, data_dir = input) {
  data <- read.delim(file.path(data_dir, paste0(label, "_curves.tsv")), check.names = FALSE)
  save_plot(paste0(label, "_posterior"), 10, 3.2, function() {
    par(mfrow = c(1, 2), mar = c(4.1, 4.4, 1.2, 1), oma = c(0, 0, 2, 0))
    for (predictor in c("mu", "log_sigma")) {
      d <- data[data$predictor == predictor, ]
      d <- d[order(d$time), ]
      noise <- predictor == "log_sigma"
      yr <- range(d$q05, d$q95, if (!noise) observed$acceleration_scaled else NULL)
      stopifnot(all(is.finite(yr)), !noise || min(yr) > 0)
      plot(d$time, d$q50, type = "n", ylim = yr, log = if (noise) "y" else "",
           xlab = "Time after impact (ms)", ylab = if (noise) "Conditional SD (scaled)" else "Acceleration (scaled)",
           panel.first = grid(col = "#EFEFEF"))
      for (pair in list(c("q05", "q95"), c("q10", "q90"), c("q25", "q75"))) {
        polygon(c(d$time, rev(d$time)), c(d[[pair[[1]]]], rev(d[[pair[[2]]]])),
                col = adjustcolor(blue, alpha.f = .23), border = NA)
      }
      lines(d$time, d$q50, col = blue, lwd = 1.6)
      if (!noise) points(observed$time, observed$acceleration_scaled, pch = 16,
                         cex = .45, col = adjustcolor("black", alpha.f = .65))
      legend("topright", if (noise) "Inferred conditional SD" else c("Inferred mean", "Observations"),
             col = if (noise) blue else c(blue, "black"),
             lty = if (noise) 1 else c(1, NA), pch = if (noise) NA else c(NA, 16),
             bty = "n", cex = .72)
    }
    mtext(title, side = 3, outer = TRUE, cex = 1.2)
  })
}
posterior("noncentered", "Noncentered pilot")

centering <- read_table("centeredness.tsv")
scatter <- function(label, geometry, title) {
  gp <- lapply(c("mu", "log_sigma"), function(p) read_table(paste0(label, "_", p, "_weights.tsv")))
  save_plot(paste0(geometry, "_scatter"), 10, 10, function() {
    par(mfrow = c(4, 4), mar = c(4.1, 4.3, 1.0, .6), oma = c(0, 0, 2, 0),
        mgp = c(2.5, .7, 0), cex = .8, mex = 1)
    for (basis in bases) {
      for (g in seq_along(gp)) {
        d <- gp[[g]][gp[[g]]$basis == basis, ]
        weights <- switch(geometry,
          centered = d$physical_weight,
          optimal = d$coordinate * exp(d$log_spectral_scale *
                     centering[centering$basis == basis, if (g == 1) "mean" else "log_scale"]),
          d$coordinate)
        for (hyper in c("sigma", "rho")) {
          plot(d[[hyper]], weights, log = "x", pch = 16, cex = .40,
               col = adjustcolor(colors[[match(basis, bases)]], alpha.f = .12),
               xlab = paste(if (g == 1) "Mean:" else "Log-SD:",
                            if (hyper == "sigma") "marginal SD" else "length scale"),
               ylab = paste(if (g == 1) "Mean weight" else "Log-SD weight", basis),
               panel.first = grid(col = "#EFEFEF"))
        }
      }
    }
    mtext(title, side = 3, outer = TRUE, cex = 1.2, font = 2)
  })
}
scatter("noncentered", "noncentered", "Noncentered pilot coordinates")
scatter("noncentered", "centered", "Centered geometry — transformed pilot draws, not another fit")
scatter("noncentered", "optimal", "Selected partial geometry — transformed pilot draws")

loss <- read_table("loss_profiles.tsv")
save_plot("loss_profiles", 10, 4.2, function() {
  par(mfrow = c(1, 2), mar = c(4, 4, 2.2, 1))
  for (gp in c("mu", "log_sigma")) {
    plot(0:1, 0:1, type = "n", xlab = "Centeredness", ylab = "Loss (each curve rescaled to [0, 1])",
         main = if (gp == "mu") "Mean GP" else "Log-SD GP")
    for (b in seq_along(bases)) {
      d <- loss[loss$predictor == gp & loss$basis == bases[[b]], ]
      ok <- is.finite(d$loss) & as.logical(d$admissible)
      yr <- range(d$loss[ok])
      values <- rep(NA_real_, nrow(d))
      values[ok] <- if (diff(yr) == 0) 0 else (d$loss[ok] - yr[[1]]) / diff(yr)
      lines(d$centeredness, values, lwd = 2, col = colors[[b]])
    }
    legend("topright", paste("Basis", bases), col = colors, lty = 1, bty = "n", cex = .8)
  }
})
save_plot("selected_centeredness", 8, 4.3, function() {
  par(mar = c(4, 4, 2, 1))
  matplot(centering$basis, centering[, c("mean", "log_scale")], type = "o",
          pch = c(16, 17), lty = 1, lwd = 2, col = colors[1:2], ylim = c(0, 1),
          xlab = "Basis frequency", ylab = "Selected centeredness", main = "One centering per basis function")
  legend("topright", c("Mean GP", "Log-SD GP"), col = colors[1:2], pch = c(16, 17), lty = 1, bty = "n")
})

if (file.exists(file.path(input, "partial_curves.tsv"))) {
  posterior("partial", "Fresh selected-partial fit")
  scatter("partial", "partial", "Selected-partial refit — newly sampled coordinates")
}
if (file.exists(file.path(online_input, "online_curves.tsv"))) {
  posterior("online", "Online adaptive centering", online_input)
  online <- read.delim(file.path(online_input, "online_centeredness.tsv"))
  save_plot("online_centeredness", 10, 3.7, function() {
    par(mfrow = c(1, 2), mar = c(4, 4, 2.3, 1))
    for (gp in c("mu", "log_sigma")) {
      d <- online[online$predictor == gp, ]
      d <- d[order(d$basis), ]
      plot(d$basis, d$centeredness, type = "o", pch = 16, lwd = 2, col = blue,
           ylim = c(0, 1), xlab = "Basis frequency", ylab = "Centeredness",
           main = if (gp == "mu") "Mean GP" else "Log-SD GP")
      lines(centering$basis, centering[[if (gp == "mu") "mean" else "log_scale"]],
            col = colors[[2]], lty = 2, lwd = 2)
      legend("bottomleft", c("Online warmup", "Offline pilot selection"),
             col = colors[1:2], lty = c(1, 2), pch = c(16, NA), bty = "n", cex = .8)
    }
  })
}

File: research/adaptive_centering/plot_results.jl
# Recreate the case-study figures from saved full-fit exports, using Julia/AoV.
# No fitting or posterior subsampling occurs here. Makie provides figure layout;
# all data marks are AlgebraOfVega specifications. HSGP coordinate changes use BRM.
using BayesianRegressionModels, AlgebraOfVega, CSV, Tables, CairoMakie

const CASE_BASES = (1, 2, 19, 20)
const CASE_COLORS = ["#0B7BEC", "#E67E22", "#16877A", "#984EA3"]
const CASE_GPS = ("mu", "log_sigma")
gp_label(p) = p == "mu" ? "Mean GP" : "Log-SD GP"
basis_label(b) = "Basis $(lpad(b, 2, '0'))"
case_table(dir, name) = collect(Tables.namedtupleiterator(
    CSV.File(joinpath(dir, name); delim='\t')))

function save_case_panels(output, name, panels; title, size=(1300, 460), subtitles=nothing)
    fig = Figure(; size, fontsize=15)
    Label(fig[0, 1:length(panels)], title; fontsize=21, font=:bold)
    for (i, panel) in enumerate(panels)
        slot = fig[1, i] = GridLayout()
        isnothing(subtitles) || Label(slot[0, 1], subtitles[i]; fontsize=18, font=:bold)
        sdraw!(slot[1, 1], panel)
    end
    path = joinpath(output, "$name.png")
    save(path, fig; px_per_unit=1.5)
    println("figure\t", path)
    path
end

function case_basis_panels()
    rows = [(; x, value=sin(pi / 3 * (x + 1.5) * j) / sqrt(1.5),
               basis=basis_label(j)) for j in CASE_BASES for x in range(-1.5, 1.5; length=501)]
    basis = (data(rows) * mapping(:x => "Scaled time", :value => "Basis function";
                 color=:basis) * visual(Lines; linewidth=1.5) +
             data((; x=[-1.0, 1.0])) * mapping(:x => "Scaled time") *
                 visual(VLines; color="#777777", linestyle=:dash)) *
        config(axis=(; title="HSGP basis functions"),
               scales=scales(Color=(; palette=CASE_COLORS)))
    # Introductory mathematical illustration, not posterior reconstruction.
    rows = [(; frequency=j,
               sd=exp(-0.25 * (j * exp(r) * pi / 3)^2 + 0.45946926660233633 + 0.5r),
               length="log length = $r") for r in (-2, -1, 0)
            for j in range(1, 20; length=201)]
    spectrum = data(rows) * mapping(:frequency => "Basis frequency",
        :sd => "Prior spectral SD"; color=:length) * visual(Lines; linewidth=2) *
        config(axis=(; title="Squared-exponential spectrum"),
               scales=scales(Color=(; palette=CASE_COLORS)))
    [basis, spectrum]
end

function case_posterior_panels(input, label, observed)
    curves = case_table(input, "$(label)_curves.tsv")
    map(CASE_GPS) do predictor
        rows = sort(filter(r -> r.predictor == predictor, curves); by=r -> r.time)
        isempty(rows) && error("Missing $predictor posterior curves")
        noise = predictor == "log_sigma"
        all(r -> all(isfinite, (r.q05, r.q50, r.q95)) && (!noise || r.q05 > 0), rows) ||
            error("Non-finite posterior interval or non-positive conditional SD")
        brm_posteriorplot(rows; xlabel="Time after impact (ms)",
            ylabel=noise ? "Conditional SD (scaled)" : "Acceleration (scaled)",
            observations=noise ? nothing : observed, observed_y=:acceleration_scaled,
            logscale=noise)
    end
end

"""Adapt exported draw tables to BRM's native frame transport; never thin draws."""
function case_pair_rows(input, label, predictor; from=zeros(20), to=from)
    rows = case_table(input, "$(label)_$(predictor)_weights.tsv")
    columns = [sort(filter(r -> r.basis == b, rows); by=r -> r.draw) for b in CASE_BASES]
    indices = getproperty.(first(columns), :draw)
    isempty(indices) && error("No saved coordinates")
    length(unique(indices)) == length(indices) || error("Duplicate saved draw IDs")
    all(col -> getproperty.(col, :draw) == indices, columns) ||
        error("Basis columns do not contain identical saved draws")
    coordinates = hcat([getproperty.(col, :coordinate) for col in columns]...)
    log_scales = hcat([getproperty.(col, :log_spectral_scale) for col in columns]...)
    transformed = hsgp_transform_draws(coordinates, log_scales;
        from=from[collect(CASE_BASES)], to=to[collect(CASE_BASES)])
    all(transformed.finite) || error("Non-finite transformed coordinates; no points were dropped")
    output = [(; basis_label=basis_label(b),
               parameter=hyper == :sigma ? "Marginal SD" : "Length scale",
               hyperparameter=getproperty(columns[j][i], hyper),
               coordinate=transformed.coordinates[i, j])
              for (j, b) in enumerate(CASE_BASES) for hyper in (:sigma, :rho)
              for i in eachindex(indices)]
    all(r -> isfinite(r.hyperparameter) && r.hyperparameter > 0, output) ||
        error("Invalid hyperparameters for logarithmic pair-plot axis")
    println("pair_data\t", label, "\t", predictor, "\tdraws_per_panel=", length(indices),
            "\trows=", length(output))
    output
end

function case_loss_rows(input)
    loss = case_table(input, "loss_profiles.tsv")
    output = NamedTuple[]
    for predictor in CASE_GPS, b in CASE_BASES
        curve = sort(filter(r -> r.predictor == predictor && r.basis == b, loss);
                     by=r -> r.centeredness)
        admissible(r) = isfinite(r.loss) && r.admissible == true
        ok = filter(admissible, curve)
        isempty(ok) && error("No admissible losses for $predictor basis $b")
        low, high = extrema(getproperty.(ok, :loss))
        segment = 1
        for r in curve
            if !admissible(r)
                segment += 1 # Never connect across an inadmissible interval.
                continue
            end
            push!(output, (; predictor=gp_label(predictor), basis_label=basis_label(b),
                segment="$predictor-$b-$segment", centeredness=r.centeredness,
                loss=high == low ? 0.0 : (r.loss - low) / (high - low)))
        end
    end
    output
end

function plot_results(input; online_input=input, output=joinpath(input, "figures"))
    mkpath(output)
    observed = case_table(input, "observations.tsv")
    selected = sort(case_table(input, "centeredness.tsv"); by=r -> r.basis)
    getproperty.(selected, :basis) == collect(1:20) || error("Expected all 20 selected basis entries")
    profiles = Dict("mu" => getproperty.(selected, :mean),
                    "log_sigma" => getproperty.(selected, :log_scale))
    paths = String[]
    push!(paths, save_case_panels(output, "hsgp_basis", case_basis_panels();
        title="HSGP basis and prior spectrum"))
    for (label, title, dir) in (("noncentered", "Noncentered pilot", input),
                               ("partial", "Fresh selected-partial fit", input),
                               ("online", "Online adaptive centering", online_input))
        isfile(joinpath(dir, "$(label)_curves.tsv")) || continue
        push!(paths, save_case_panels(output, "$(label)_posterior",
            case_posterior_panels(dir, label, observed); title))
    end
    for (label, geometry, title) in (
            ("noncentered", "noncentered", "Noncentered pilot coordinates"),
            ("noncentered", "centered", "Centered geometry — transformed pilot draws, not another fit"),
            ("noncentered", "optimal", "Selected partial geometry — transformed pilot draws"),
            ("partial", "partial", "Selected-partial refit — newly sampled coordinates"))
        isfile(joinpath(input, "$(label)_mu_weights.tsv")) || continue
        panels = map(CASE_GPS) do predictor
            from = label == "partial" ? profiles[predictor] : zeros(20)
            to = geometry == "centered" ? ones(20) :
                 geometry == "optimal" ? profiles[predictor] : from
            brm_pairplot(case_pair_rows(input, label, predictor; from, to))
        end
        push!(paths, save_case_panels(output, "$(geometry)_scatter", panels;
            title, size=(1500, 1350), subtitles=gp_label.(CASE_GPS)))
    end
    push!(paths, save_case_panels(output, "loss_profiles",
        [brm_centering_lossplot(case_loss_rows(input);
             ylabel="Offline loss (each curve rescaled to [0, 1])")];
        title="Post-hoc pilot selection — offline objective"))
    centeredness = [(; basis=r.basis, predictor=gp_label(p),
                      centeredness=profiles[p][r.basis]) for p in CASE_GPS for r in selected]
    push!(paths, save_case_panels(output, "selected_centeredness",
        [brm_centerednessplot(centeredness)]; title="One centering per basis function",
        size=(1000, 470)))
    if isfile(joinpath(online_input, "online_centeredness.tsv"))
        online = case_table(online_input, "online_centeredness.tsv")
        learned = Dict(p => getproperty.(sort(filter(r -> r.predictor == p, online);
                                             by=r -> r.basis), :centeredness) for p in CASE_GPS)
        all(p -> length(learned[p]) == 20, CASE_GPS) || error("Incomplete online centering profile")
        compared = vcat([merge(r, (; configuration="Post-hoc pilot")) for r in centeredness],
            [(; basis=r.basis, predictor=gp_label(r.predictor), centeredness=r.centeredness,
                configuration="Online warmup") for r in online])
        push!(paths, save_case_panels(output, "online_centeredness",
            [brm_centerednessplot(compared; compare=true)];
            title="Online and post-hoc selected centeredness"))
        panels = map(CASE_GPS) do p
            # WarmupHMC returns final draws in the original NCP frame.
            brm_pairplot(case_pair_rows(online_input, "online", p; to=learned[p]))
        end
        push!(paths, save_case_panels(output, "online_scatter", panels;
            title="Online fit — coordinates in learned geometry", size=(1500, 1350),
            subtitles=gp_label.(CASE_GPS)))
    end
    println("render_complete\tfigures=", length(paths), "\tbackend=AlgebraOfVega/CairoMakie")
    paths
end

if abspath(PROGRAM_FILE) == @__FILE__
    1 <= length(ARGS) <= 3 || error("Usage: plot_results.jl OFFLINE_DIR [ONLINE_DIR [OUTPUT_DIR]]")
    plot_results(ARGS[1]; online_input=length(ARGS) >= 2 ? ARGS[2] : ARGS[1],
                 output=length(ARGS) == 3 ? ARGS[3] : joinpath(ARGS[1], "figures"))
end
